pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java

```diff
@@ -1280,12 +1279,15 @@ private void failPendingMessages(ClientCnx cnx, PulsarClientException ex) {
                 ReferenceCountUtil.safeRelease(op.cmd);
                 op.recycle();
             });
-            semaphore.release(releaseCount.get());
+
             pendingMessages.clear();
             pendingCallbacks.clear();
             if (isBatchMessagingEnabled()) {
                 failPendingBatchMessages(ex);
             }
+
+            semaphore.drainPermits();
+            semaphore.release(conf.getMaxPendingMessages());
```

Comment:
This doesn't look correct to me, what's the purpose of release 1000 permits on the semaphore?

Reply:
You are right, I have fixed it.